[Wen Xuan] iP

wxliong opened the pull request:
# wHy sO SeRioUS
> “I wear a mask. And that mask, it’s not to hide who I am, but to create what I am.” - Batman telling you to wear your masks during the pandemic

Batman (but mostly Alfred, his butler) will help to handle your tasks in a neat and intuitive way 👴🏻 

**Features:**
- [X] Greeting & Exiting 🦇
- [X] Adding tasks (Todo, Event, Deadline) 
- [X] Changing tasks status ✅
- [X] ~Deleting tasks~ 🗑
- [X] Listing all tasks 📝
- [ ] Whatever you desire 👹

**How to start?**
1. Make sure you're using [Java 11](https://www.oracle.com/java/technologies/downloads/#java11)
2. Download batman from [here](https://github.com/wxliong/ip/releases/download/A-Jar/Duke.jar)
3. Channel your inner batman by running: `java -jar Duke.jar`
4. Let Alfred handle your life for you 🙂

**Insert random code block here:**
```java
boolean batman = false;
if (batman) {
    System.out.println("Congrats, you're BATMAN!");
} else {
    System.out.println("Congrats, you're a BUTLER!");
}
```
###### Note: If anything goes wrong, remember Alfred is the one operating things. _Do you really want to blame an old man?_

Commit:
Add Gradle support

wxliong committed:
Added Greet, Echo, Exit to Duke.java

wxliong committed:
Add "Add" and "List" to Duke.java

wxliong committed:
Add the ability to mark/unmark tasks

wxliong committed:
Add Todo, Events, Deadlines

- ToDos: tasks without any date/time attached to it e.g., visit new theme park
- Deadlines: tasks that need to be done before a specific date/time
- Events: tasks that start at a specific time and ends at a specific time

wxliong committed:
Add Automated Text UI Testing

- Remove extra newlines from Duke.java
- Change unreadable characters to whitespaces

wxliong committed:
Add DukeException.java

- Update Duke.java to handle errors

wxliong committed:
Add the ability to delete tasks from list

- Update error handling for deletion

wxliong committed:
Add Enums

- Enumerations add for different exceptions
e.g. empty desc, invalid command

wxliong committed:
Add ability to save and load tasks from file
- Add new class FileSave to handle file capabilities
- Update exceptions to handle file errors

wxliong committed:
Add DateUtil to handle date/time formats
- Update Exceptions to handle wrong date/time formats
- Update test cases for date/time formats

wxliong committed:
Merge branch 'branch-Level-7'

wxliong committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	.gitignore
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

wxliong committed:
Make the code more OOP
- Add UI class to deal with interactions with the user
- Add Storage class to deal with loading/saving tasks in file
- Add Parser class to deal with user commands
- Add TaskList class to handle operations relating to tasks

wxliong committed:
Divide classes into packages

wxliong committed:
Add JUnit tests

wxliong committed:
Package the App as a JAR file

wxliong committed:
Tweak the code to comply with a coding standard
- Add files for checkstyle

wxliong committed:
Add JavaDoc comments

wxliong committed:
Add the ability to find a task using keyword
- Update testcases according to the find function

wxliong committed:
Merge branch 'branch-Level-9'

wxliong committed:
Merge branch 'branch-A-CodingStandard'

wxliong committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/batman/tasks/Deadline.java
#	src/main/java/batman/tasks/Event.java
#	src/main/java/batman/tasks/TaskList.java

wxliong committed:
Fix checkstyle violations

wxliong committed:
Merge branch 'add-gradle-support' of https://github.com/wxliong/ip into branch-A-Gradle

wxliong committed:
Automate project builds using Gradle

wxliong committed:
Merge branch 'branch-A-Gradle'

wxliong committed:
Fix coding style violations

wxliong committed:
Revert "Fix coding style violations"

This reverts commit a7987a1d39fdb9d2299a0e4d9feef1315654dde5.

wxliong committed:
Fix coding style violations

wxliong committed:
Merge branch 'branch-A-CheckStyle'

JonathanHoshi commented on src/main/java/batman/tasks/Task.java:
```diff
@@ -0,0 +1,39 @@
+        return "Now you have " + totalTasks + " tasks in the list.";
+    }
+
+    public abstract String appendtoFile();
```
appendtoFile() should be in camelCase -> appendToFile()

JonathanHoshi commented on src/main/java/batman/tasks/Event.java:
```diff
@@ -0,0 +1,45 @@
+    protected LocalDateTime at;
+
+    /**
+     * @param isDone Whether the task is done.
```
Shouldn't there be a method description in the JavaDoc comments for all your methods? Good job for your params, throws and return sections of the JavaDoc.

JonathanHoshi commented on src/main/java/batman/exception/Error.java:
```diff
@@ -0,0 +1,7 @@
+
+public enum Error {
+    INVALID,
+    EMPTYDESC,
+    LISTERROR
```
Shouldn't your constants name be all uppercase using underscore to separate words? -> EMPTY_DESC and LIST_ERROR

JonathanHoshi commented on src/main/java/batman/exception/DukeException.java:
```diff
@@ -0,0 +1,26 @@
+
+    /**
+     * Returns a string if an invalid input is encountered.
+     */
```
Shouldn't there be a @return parameter in the JavaDoc comment?

JonathanHoshi commented on src/main/java/batman/tasks/TaskList.java:
```diff
@@ -0,0 +1,186 @@
+     *
+     * @param command An array of strings containing the
+     *                details of a task.
+     */
```
Shouldn't there be a @return parameter in the JavaDoc comments? 
```suggestion
     @return the response message.
     */
```

JonathanHoshi reviewed (commented):
Overall, I feel that you have follow the Java coding standards pretty well. I have also noted some minor coding standard violations. Good Job !👍

wxliong committed:
Add a GUI to Duke

wxliong committed:
Merge branch 'branch-Level-10'

wxliong committed:
Fix GUI bug

wxliong committed:
Update gradle file

wxliong committed:
Fix exit feature and coding standard violation
- Exit the program after user reply with a "bye"
- Fix weak SLAP in class "TaskList"

wxliong committed:
Add assertions

Assertions are used to verify the correctness of an invariant in the
code.

Let's,
- use assertions to check if nulls exists
- use assertions to check if task file exists or not

wxliong committed:
Merge pull request #2 from wxliong/branch-A-Assertions

Add assertions

wxliong committed:
Improve code quality

- Fix javadocs
- Remove unused variable

wxliong committed:
Revert "Improve code quality"

This reverts commit 9815bb326dceeceb12b38fb796c8e4f30dd063ef.

wxliong committed:
Improve code quality

- Fix javadocs
- Remove unused variable

wxliong committed:
Merge pull request #3 from wxliong/branch-A-CodeQuality

Improve code quality

wxliong committed:
Add streams to code

wxliong committed:
Merge pull request #4 from wxliong/branch-A-Streams

Add streams to code

wxliong committed:
Add sorting to tasks with timeline
- Event, Deadline will be sorted accordingly to dates.
- Todo tasks will be above Event and Deadline.
- An assumption is that the dates given to Event and Deadline is smaller
than Todo.

wxliong committed:
Merge branch 'branch-BCD-Extension'

wxliong committed:
Update GUI
- Add background
- Add solid colours to dialogbox and icons

wxliong committed:
Add a User Guide

wxliong committed:
Update README.md

wxliong committed:
Update README.md

wxliong committed:
Update README.md
